src/Internal/Utils.elm

```diff
@@ -0,0 +1,197 @@
+                        -- We can just save the rest of Left.
+                        let
+                            newTree =
+                                treeZipper
+                                    |> TreeZipper.tree
+                                    |> Tree.prependChild (Tree.singleton splitInfo.left)
+                        in
+                        treeZipper
+                            |> TreeZipper.replaceTree newTree
+
+                    else
+                        let
+                            maybeNextTreeZipper =
+                                treeZipper
+                                    -- Try to advance
+                                    |> TreeZipper.forward
+                        in
+                        case maybeNextTreeZipper of
+                            Just nextTreeZipper ->
+                                -- Recurse.
+                                insertHelper funcs splitInfo.left nextTreeZipper
+
+                            Nothing ->
+                                treeZipper
+
+                else
+                    -- Nothing remains to be added.
+                    treeZipper
+
+            else
+                -- We have a partial match with the existing tree. We need to add a new
+                -- parent tree.
+                let
+                    rightTreeUpdated =
+                        treeZipper
+                            -- Keep only the diff elements of the Right element.
+                            |> TreeZipper.replaceLabel splitInfo.right
+                            -- Get the new tree from the focus point.
+                            |> TreeZipper.tree
+
+                    newTree =
+                        Tree.singleton splitInfo.intersect
+                            |> Tree.prependChild rightTreeUpdated
+                            |> Tree.prependChild (Tree.singleton splitInfo.left)
+                in
+                treeZipper
+                    |> TreeZipper.replaceTree newTree
+
+
+{-| Return a record holding the matching elements, the ones the differ from the "left"
+(i.e. new items being inserted), and the ones on the "right" (i.e. out of an existing
+node of the tree).
+-}
+split : { intersectFunc : List a -> List a -> List a, removeFunc : List a -> List a -> List a } -> List a -> List a -> { intersect : List a, left : List a, right : List a }
+split { intersectFunc, removeFunc } xs ys =
+    let
+        intersect =
+            intersectFunc xs ys
+
+        left =
+            removeFunc xs intersect
+
+        right =
+            List.foldl (\x accum -> List.Extra.remove x accum) ys intersect
+    in
+    { intersect = intersect
+    , left = left
+    , right = right
+    }
+
+
+{-| Find the matching elements, regardless of their order in the list.
+-}
+unOrderedIntersect : List a -> List a -> List a
+unOrderedIntersect xs ys =
+    List.foldr
```

Comment:
If the order doesn't matter, then `List.foldl` is much faster.

Reply:
It matters for the order. I can add a comment